src/sqlfluff/utils/reflow/reindent.py

```diff
@@ -27,3 +36,792 @@ def deduce_line_indent(raw_segment: RawSegment, root_segment: BaseSegment) -> st
+            # NOTE: Last edge case. If we haven't yielded yet, but the
+            # next element is the end of the file. Yield.
+            elif (
+                indent_impulse
+                or indent_trough
+                or idx == 0
+                or elements[idx + 1].segments[0].is_type("end_of_file")
+            ):
+                yield _IndentPoint(
+                    idx,
+                    indent_impulse,
+                    indent_trough,
+                    indent_balance,
+                    last_line_break_idx,
+                    False,
+                    untaken_indents,
+                )
+                has_newline = False
+
+            # Strip any untaken indents above the new balance.
+            # NOTE: We strip back to the trough, not just the end point
+            # if the trough was lower than the impulse.
+            untaken_indents = tuple(
+                x
+                for x in untaken_indents
+                if x
+                <= (
+                    indent_balance + indent_impulse + indent_trough
+                    if indent_trough < indent_impulse
+                    else indent_balance + indent_impulse
+                )
+            )
+
+            # After stripping, we may have to add them back in.
+            if indent_impulse > indent_trough and not has_newline:
+                for i in range(indent_trough, indent_impulse):
+                    untaken_indents += (indent_balance + i + 1,)
+
+            # Update values
+            indent_balance += indent_impulse
+
+
+def _map_line_buffers(elements: ReflowSequenceType) -> List[_IndentLine]:
+    """Map the existing elements, building up a list of _IndentLine."""
+    # First build up the buffer of lines.
+    lines = []
+    point_buffer = []
+    for indent_point in _crawl_indent_points(elements):
+        # We evaluate all the points in a line at the same time, so
+        # we first build up a buffer.
+        point_buffer.append(indent_point)
+
+        if not indent_point.is_line_break:
+            continue
+
+        # If it *is* a line break, then store it.
+        lines.append(_IndentLine.from_points(point_buffer))
+        # Reset the buffer
+        point_buffer = [indent_point]
+
+    # Handle potential final line
+    if len(point_buffer) > 1:
+        lines.append(_IndentLine.from_points(point_buffer))
+
+    return lines
+
+
+def _evaluate_indent_point_buffer(
+    elements: ReflowSequenceType,
+    indent_line: _IndentLine,
+    single_indent: str,
+    forced_indents: List[int],
+) -> List[LintFix]:
+    """Evalute a single set of indent points on one line.
+
+    NOTE: This mutates the given `elements` and `forced_indents` input to avoid
+    lots of copying.
+
+    Order of operations:
+    1. Evaluate the starting indent for this line.
+    2. For point which aren't line breaks in the line, we evaluate them
+       to see whether they *should* be. We seperately to missing indents
```

Comment:
This reads a little funny -- can you review for correctness?
> We seperately to missing indents

Reply:
This whole point is badly worded. Thanks for catching it - I think I probably wrote this while everything wasn't quite working yet. 😵